File: 2024-10-October-GFG-POTD/October 7 - XOR Linked list/Solution.java
import java.io.*;
import java.util.*;

class Node {
    int data;
    Node npx; // XOR of next and previous node

    Node(int data) {
        this.data = data;
        this.npx = null;
    }
}

class XORLinkedList {
    // Utility function to XOR two Node references
    static Node XOR(Node a, Node b) {
        return (Node)((a != null ? a.hashCode() : 0) ^ (b != null ? b.hashCode() : 0));
    }

    // Function to insert a new node at the front of the XOR LinkedList
    static Node insert(Node head, int data) {
        Node newNode = new Node(data);
        newNode.npx = head;

        // If the list is not empty, adjust the previous head's npx
        if (head != null) {
            head.npx = XOR(newNode, head.npx);
        }
        return newNode; // New node becomes the head
    }

    // Function to get the list in forward order
    static List<Integer> getList(Node head) {
        List<Integer> result = new ArrayList<>();
        Node curr = head, prev = null, next;

        // Traverse the XOR linked list
        while (curr != null) {
            result.add(curr.data);
            // Update next node
            next = XOR(prev, curr.npx);
            prev = curr;
            curr = next;
        }
        return result;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(br.readLine());

        while (t-- > 0) {
            Node head = null;
            String input = br.readLine();
            StringTokenizer st = new StringTokenizer(input);

            while (st.hasMoreTokens()) {
                int number = Integer.parseInt(st.nextToken());
                head = insert(head, number);
            }

            List<Integer> list = getList(head);
            // Print list in forward order
            for (int x : list) {
                System.out.print(x + " ");
            }
            System.out.println();

            // Print list in reverse order
            for (int i = list.size() - 1; i >= 0; i--) {
                System.out.print(list.get(i) + " ");
            }
            System.out.println();
        }
    }
}

File: 2024-10-October-GFG-POTD/October 7 - XOR Linked list/Solution.py
class Node:
    def __init__(self, data):
        self.data = data
        self.npx = None  # XOR of next and previous node

# Utility function to XOR two node references
def XOR(a, b):
    return a if not b else b if not a else a ^ b

# Function to insert a new node at the front of the XOR LinkedList
def insert(head, data):
    new_node = Node(data)
    new_node.npx = id(head)

    # If the list is not empty, adjust the previous head's npx
    if head:
        head.npx = XOR(new_node, head.npx)
    return new_node  # New node becomes the head

# Function to get the list in forward order
def get_list(head):
    result = []
    curr, prev, next_node = head, None, None

    while curr:
        result.append(curr.data)
        next_node = XOR(prev, curr.npx)
        prev, curr = curr, next_node

    return result

def main():
    t = int(input())

    for _ in range(t):
        head = None
        numbers = map(int, input().split())

        for number in numbers:
            head = insert(head, number)

        result = get_list(head)

        # Print list in forward order
        print(" ".join(map(str, result)))

        # Print list in reverse order
        print(" ".join(map(str, reversed(result))))

if __name__ == "__main__":
    main()
